File: leetcode_rust/src/kadanes_algo.rs
/*
Given an integer array nums, find the
subarray
with the largest sum, and return its sum.

*/


use std::cmp::max;

impl Solution {
    pub fn max_sub_array(nums: Vec<i32>) -> i32 {
        nums.iter().fold((i32::MIN, 0), |(maxsum, cursum), &num| {
            let cursum = max(cursum + num, num);
            (max(maxsum, cursum), cursum)
        }).0
    }
}

File: leetcode/kadanes_algo.cpp
/*
Given an integer array nums, find the
subarray
with the largest sum, and return its sum.
 */

class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int maxsum = numeric_limits<int>::min();
        int currsum(0);

        for(int nn : nums){
            currsum += nn;
            maxsum = max(currsum, maxsum);
            if(currsum<0) currsum = 0;
        }

        return maxsum;
        
    }
};
